source/hedgehog/Main.hs

```diff
@@ -0,0 +1,419 @@
+  property "Int32" . tryFromTryFrom @Int.Int32 $ gen 0 2147483647
+  property "Int64" . tryFromTryFrom @Int.Int64 $ gen 0 9223372036854775807
+  property "Int" . tryFromTryFrom @Int $ gen 0 9223372036854775807
+  property "Integer" . fromTryFrom @Integer $ gen 0 18446744073709551615
+  property "Word8" . tryFromFrom @Word.Word8 $ gen 0 255
+  property "Word16" . tryFromFrom @Word.Word16 $ gen 0 65535
+  property "Word32" . tryFromTryFrom @Word.Word32 $ gen 0 4294967295
+  property "Word64" . fromTryFrom @Word.Word64 $ gen 0 18446744073709551615
+  property "Natural" . fromTryFrom @Natural.Natural $ gen 0 18446744073709551615
+  property "Float" . tryFromTryFrom @Float $ gen 0 16777215
+  property "Double" . tryFromTryFrom @Double $ gen 0 9007199254740991
+
+groupNatural :: H.Group
+groupNatural = group "Natural" $ do
+  let gen lo hi = Gen.integral $ Range.linear lo hi :: H.Gen Natural.Natural
+  property "Int8" . tryFromTryFrom @Int.Int8 $ gen 0 127
+  property "Int16" . tryFromTryFrom @Int.Int16 $ gen 0 32767
+  property "Int32" . tryFromTryFrom @Int.Int32 $ gen 0 2147483647
+  property "Int64" . tryFromTryFrom @Int.Int64 $ gen 0 9223372036854775807
+  property "Int" . tryFromTryFrom @Int $ gen 0 9223372036854775807
+  property "Integer" . fromTryFrom @Integer $ gen 0 99999999999999999999
+  property "Word8" . tryFromFrom @Word.Word8 $ gen 0 255
+  property "Word16" . tryFromFrom @Word.Word16 $ gen 0 65535
+  property "Word32" . tryFromFrom @Word.Word32 $ gen 0 4294967295
+  property "Word64" . tryFromFrom @Word.Word64 $ gen 0 18446744073709551615
+  property "Word" . tryFromFrom @Word $ gen 0 18446744073709551615
+  property "Float" . tryFromTryFrom @Float $ gen 0 16777215
+  property "Double" . tryFromTryFrom @Double $ gen 0 9007199254740991
+
+groupFloat :: H.Group
+groupFloat = group "Float" $ do
+  let genI lo hi = fmap (fromIntegral @Int.Int32) . Gen.integral $ Range.linear lo hi :: H.Gen Float
+  property "Int8" . tryFromFrom @Int.Int8 $ genI -128 127
+  property "Int16" . tryFromFrom @Int.Int16 $ genI -32768 32767
+  property "Int32" . tryFromTryFrom @Int.Int32 $ genI -16777215 16777215
+  property "Int64" . tryFromTryFrom @Int.Int64 $ genI -16777215 16777215
+  property "Int" . tryFromTryFrom @Int $ genI -16777215 16777215
+  property "Integer" . tryFromTryFrom @Integer $ genI -16777215 16777215
+  property "Word8" . tryFromFrom @Word.Word8 $ genI 0 255
+  property "Word16" . tryFromFrom @Word.Word16 $ genI 0 65535
+  property "Word32" . tryFromTryFrom @Word.Word32 $ genI 0 16777215
+  property "Word64" . tryFromTryFrom @Word.Word64 $ genI 0 16777215
+  property "Word" . tryFromTryFrom @Word $ genI 0 16777215
+  property "Natural" . tryFromTryFrom @Natural.Natural $ genI 0 16777215
+  let genF lo hi = Gen.realFloat $ Range.linearFrac lo hi :: H.Gen Float
+  property "Rational" . tryFromFrom @Rational $ genF -16777215 16777215
+  property "Double" . fromFrom @Double $ genF -16777215 16777215
```

Comment:
This only tests "good" floats. I would have to write properties with overloaded equality in order to test "bad" floats like NaN. It might be good to explicitly test infinities and denormals though.

Reply:
Never mind, those cases can be covered by unit tests.

https://github.com/tfausak/witch/blob/39f86efb85abb28b2eeaebe1df2be0b236be5e98/source/test-suite/Main.hs#L1452